internal/controller/controllers/bmo_utils.go

```diff
@@ -0,0 +1,135 @@
+)
+
+type BMOUtils struct {
+	client.Client
+	log            logrus.FieldLogger
+	kubeAPIEnabled bool
+}
+
+func NewBMOUtils(client client.Client, log logrus.FieldLogger, kubeAPIEnabled bool) *BMOUtils {
+	return &BMOUtils{client, log, kubeAPIEnabled}
+}
+
+// ConvergedFlowAvailable checks the baremetal operator version and returns true if it's equal or higher than the minimal version for converged flow
+func (r *BMOUtils) ConvergedFlowAvailable(minimalVersionForConvergedFlow string) bool {
+	key := types.NamespacedName{
+		Name: "baremetal",
+	}
+	clusterOperator := &v1.ClusterOperator{}
+	if err := r.Client.Get(context.TODO(), key, clusterOperator); err != nil {
+		r.log.Errorf("Error querying api for baremetal operator status: %s", err)
+		return false
+	}
+	if len(clusterOperator.Status.Versions) == 0 {
+		r.log.Infof("no version found for baremetal operator")
+		return false
+	}
+	version := clusterOperator.Status.Versions[0].Version
+	r.log.Infof("The baremetal operator version is %s, the minimal version for the converged flow is %s", version, minimalVersionForConvergedFlow)
+
+	available := semver.New(version).Compare(*semver.New(minimalVersionForConvergedFlow)) >= 0
+	if available {
+		r.log.Infof("Converged flow enabled")
+	}
+	return available
+}
+
+// TODO: replace this with public function in BMO https://github.com/openshift/cluster-baremetal-operator/pull/261
+func (r *BMOUtils) GetIronicServiceURL() (string, error) {
+	provisioningInfo, err := r.readProvisioningCR()
+	if err != nil {
+		r.log.WithError(err).Error("unable to get provisioning CR")
+		return "", err
+	}
+	ironicIP, err := r.getIronicIP(*provisioningInfo)
+	if err != nil || ironicIP == "" {
+		r.log.WithError(err).Error("unable to determine Ironic's IP")
+		return "", err
+	}
+	return getUrlFromIP(ironicIP), nil
+}
+
+func (r *BMOUtils) readProvisioningCR() (*metal3iov1alpha1.Provisioning, error) {
+	// Fetch the Provisioning instance
+	instance := &metal3iov1alpha1.Provisioning{}
+	namespacedName := types.NamespacedName{Name: metal3iov1alpha1.ProvisioningSingletonName, Namespace: ""}
+	if err := r.Client.Get(context.TODO(), namespacedName, instance); err != nil {
+		return nil, errors.Wrap(err, "unable to read Provisioning CR")
+	}
+	return instance, nil
+}
+
+func (r *BMOUtils) getIronicIP(info metal3iov1alpha1.Provisioning) (string, error) {
+	// this is how BMO get the IronicIP
+	config := info.Spec
+	if config.ProvisioningNetwork != metal3iov1alpha1.ProvisioningNetworkDisabled && !config.VirtualMediaViaExternalNetwork {
+		return config.ProvisioningIP, nil
+	}
+	return r.getPodHostIP(info.Namespace)
+}
+
+func getUrlFromIP(ipAddr string) string {
+	if strings.Contains(ipAddr, ":") {
+		// This is an IPv6 addr
+		return "https://" + fmt.Sprintf("[%s]", ipAddr)
+	}
+	if ipAddr != "" {
+		// This is an IPv4 addr
+		return "https://" + ipAddr
+	} else {
+		return ""
+	}
+}
+
+func (r *BMOUtils) getPodHostIP(targetNamespace string) (string, error) {
```

Comment:
1. This is the fallback (in case the ProvisioningNetworkDisabled)
2. This is what CBO is doing (I copied this logic from CBO until they make the `GetIronicServiceURL` public).
This code will get deleted in my next PR (the function is already public but adding cluster-baremetal-operator provisioning package changed 1000+ files so I prefer separating it from this PR)
https://github.com/openshift/assisted-service/pull/3815/files#diff-2ca509f8d93cc09b0a0b0185852cf7058f932d306ee3419f345057fdddd6e168R55

Reply:
https://github.com/openshift/cluster-baremetal-operator/pull/208#issue-1024478581
I'll handle it in another PR